app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java

```diff
@@ -237,12 +244,90 @@ public boolean handleExternalUrl(final String url) {
+
+                AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
                 if (hitTarget.isLink) {
-                    final Intent shareIntent = new Intent(Intent.ACTION_SEND);
-                    shareIntent.setType("text/plain");
-                    shareIntent.putExtra(Intent.EXTRA_TEXT, hitTarget.linkURL);
-                    startActivity(shareIntent);
+                    builder.setTitle(hitTarget.linkURL);
+                } else if (hitTarget.isImage) {
+                    builder.setTitle(hitTarget.imageURL);
                 }
+
+                final NavigationView menu = new NavigationView(getContext());
+                menu.inflateMenu(R.menu.menu_browser_context);
+
+                menu.getMenu().findItem(R.id.menu_link_share).setEnabled(hitTarget.isLink);
+                menu.getMenu().findItem(R.id.menu_link_copy).setEnabled(hitTarget.isLink);
+                menu.getMenu().findItem(R.id.menu_image_share).setEnabled(hitTarget.isImage);
+                menu.getMenu().findItem(R.id.menu_image_copy).setEnabled(hitTarget.isImage);
+
+                builder.setView(menu);
+
+                builder.setOnCancelListener(new DialogInterface.OnCancelListener() {
+                    @Override
+                    public void onCancel(DialogInterface dialog) {
+                        // This even is only sent when the back button is pressed, or when a user
+                        // taps outside of the dialog:
+                        TelemetryWrapper.cancelWebContextMenuEvent();
+                    }
+                });
+
+                final Dialog dialog = builder.create();
+
+                menu.setNavigationItemSelectedListener(new NavigationView.OnNavigationItemSelectedListener() {
+                    @Override
+                    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
+                        dialog.dismiss();
+
+                        switch (item.getItemId()) {
+                            case R.id.menu_link_share: {
+                                TelemetryWrapper.shareLinkEvent();
+                                final Intent shareIntent = new Intent(Intent.ACTION_SEND);
+                                shareIntent.setType("text/plain");
+                                shareIntent.putExtra(Intent.EXTRA_TEXT, hitTarget.linkURL);
+                                startActivity(shareIntent);
+                                return true;
+                            }
+                            case R.id.menu_image_share: {
+                                TelemetryWrapper.shareImageEvent();
+                                final Intent shareIntent = new Intent(Intent.ACTION_SEND);
+                                shareIntent.setType("text/plain");
+                                shareIntent.putExtra(Intent.EXTRA_TEXT, hitTarget.linkURL);
+                                startActivity(shareIntent);
```

Comment:
Make sure to create an IntentChooser Intent here (to show additional in-app share targets and to avoid launching a default for this intent automatically).